app/master/build.py

```diff
@@ -119,9 +120,13 @@ def prepare(self, subjobs, job_config):
         self._unstarted_subjobs = Queue(maxsize=len(subjobs))
         self._finished_subjobs = Queue(maxsize=len(subjobs))
 
-        for subjob in subjobs:
-            self._all_subjobs_by_id[subjob.subjob_id()] = subjob
-            self._unstarted_subjobs.put(subjob)
+        with self._subjob_queuing_lock:
+            if self._status() in [BuildStatus.FINISHED, BuildStatus.ERROR, BuildStatus.CANCELED]:
+                self._logger.warn('Could not complete build preparation, build status is {}', self._status())
+                return
+            for subjob in subjobs:
+                self._all_subjobs_by_id[subjob.subjob_id()] = subjob
+                self._unstarted_subjobs.put(subjob)
```

Comment:
Did you run into a situation where you needed this lock? I was thinking that the line `if not self._preparation_coin.spend():` protected against this being executed more than once (implying that this couldn't be executed by multiple threads simultaneously) but I may not be remembering correctly.